### PR TITLE
Introduce `shared` package

### PR DESCRIPTION
**Pull request by wochinge**
**Proposed changes**:
- introduce the `rasa.shared` package. This package will contain items which are shared with downstream dependencies such as Rasa X **and** are specific to the Rasa Open Source domain. This means this package is supposed to include key representations such as `Event` or `TrainingData`, but ideally no helper functions (e.g. constants for the database connection). Utilities can be included in `rasa.shared` in case the `rasa.shared` package depends on these itself. Utilities shouldn't be included just because Rasa X might need them.

- fix https://github.com/RasaHQ/rasa-x/issues/3465
- fix https://github.com/RasaHQ/rasa-x/issues/3470
- add a unit test which ensures that the `rasa.shared` package is independent. `independent` means that `rasa.shared` only imports things from external libraries or `rasa.shared`, but not from other rasa packages such as `rasa.core`. This ensures that the `shared` package can easily moved to a separate library and the code is actually de-coupled.

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by m-vdb** [comment]
@wochinge one thought I had and maybe you already thought about this:
- how hard would it be to have a `rasa_shared` package (next to `rasa`, same git repo) instead of `rasa.shared`? would that be supported by `poetry` / `pip`?
- would that help us more with respect to decoupling?

**Comment by wochinge** [comment]
> would that be supported by poetry / pip?

Yes, [it's supported](https://python-poetry.org/docs/versions/#path-dependencies). Let me think about the implications :-)

**Comment by m-vdb** [comment]
for reference, discussed in meeting: we'll keep it like that (`rasa.shared`) and setup `CODEOWNERS` so that we're a bit more defensive

**Comment by wochinge** [comment]
+ we merge each iteration directly to `master`
+ the `shared` subpackage should be completely independent of the rest of Rasa Open Source (I'll set up a check for that)

File: rasa/nlu/classifiers/diet_classifier.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 
 from typing import Any, Dict, List, Optional, Text, Tuple, Union, Type, NamedTuple
 
+import rasa.shared.utils.io
```
Comment (alwx) [inline rasa/nlu/classifiers/diet_classifier.py:14]: but not the same here! :)

File: rasa/shared/core/slots.py [hunk 2]
```diff
@@ -1,10 +1,10 @@
 import logging
-from typing import Any, Dict, List, NoReturn, Optional, Text, Type
 
-from rasa.core import utils
 from typing import Any, Dict, List, NoReturn, Optional, Text, Type
-from rasa.core.constants import DEFAULT_CATEGORICAL_SLOT_VALUE
-from rasa.utils.common import class_from_module_path, raise_warning
+
+import rasa.shared.utils.common
+from rasa.shared.core.constants import DEFAULT_CATEGORICAL_SLOT_VALUE
```
Comment (alwx) [inline rasa/shared/core/slots.py:6]: again, I find it better to import the constants module itself, not separate constants

File: rasa/shared/core/slots.py [hunk 3]
```diff
@@ -1,10 +1,10 @@
...
+from rasa.shared.utils import io as io_utils, common as common_utils
```
Comment (alwx) [inline rasa/shared/core/slots.py:7]: well, I think we should choose one name for imported stuff, and use either `import rasa.shared.utils.io as io_utils` or just `import rasa.shared.utils.io`, but not both ways mixed together. 
the reason is that different synonyms for imported modules make it more difficult to search for specific functions and their usage using utils like `grep`

Comment (alwx) [reply rasa/shared/core/slots.py:7]: in addition, you import `rasa.shared.utils.common` two times here

line 5:
`import rasa.shared.utils.common`

line 7:
`from rasa.shared.utils import io as io_utils, common as common_utils`

File: rasa/train.py [hunk 4]
```diff
@@ -12,13 +12,8 @@
...
 from rasa.utils.common import TempDirectoryPath
 
-from rasa.cli.utils import (
-    print_success,
-    print_warning,
-    print_error,
-    bcolors,
-    print_color,
-)
+from rasa.cli.utils import print_success, print_warning, print_error, print_color
```
Comment (alwx) [inline rasa/train.py:15]: again, functions are imported by their names while the module itself should be imported

**Review by alwx — CHANGES_REQUESTED** [review]

File: rasa/importers/multi_project.py [hunk 5]
```diff
@@ -10,7 +10,8 @@
 from rasa.importers import utils
 from rasa.nlu.training_data import TrainingData
 from rasa.core.training.structures import StoryGraph
-from rasa.utils.common import raise_warning, mark_as_experimental_feature
+from rasa.utils.common import mark_as_experimental_feature
+from rasa.shared.utils.io import raise_warning
```
[the remark this answers is not in the document]
Comment (wochinge) [reply rasa/importers/multi_project.py:14]: Will do - I skipped it to keep the diff smaller, but I agree that it's inconsistent 👍

File: rasa/shared/utils/common.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
+from typing import Text, Optional, Any, List
+
+
+def class_from_module_path(
+    module_path: Text, lookup_path: Optional[Text] = None
+) -> Any:
+    """Given the module name and path of a class, tries to retrieve the class.
+
+    The loaded class can be used to instantiate new objects. """
+    import importlib
```
Comment (m-vdb) [inline rasa/shared/utils/common.py:10]: can we move it to the top-level?

File: tests/shared/test_shared.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+                line
+                for line in lines
+                if line.startswith("import ") or line.startswith("from ")
+            ]
+            rasa_imports = [line for line in imports if "rasa" in line]
+
+            outside_rasa_imports = [
+                import_line
+                for import_line in rasa_imports
+                if "rasa.shared" not in import_line
```
Comment (m-vdb) [inline tests/shared/test_shared.py:26]: It wouldn't work if you did something like `from rasa import shared`, but that's admittedly a detail/edge case that we can accept.

File: tests/shared/utils/test_common.py [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import random
+
+import rasa.shared.utils.common
+
+
+def test_all_subclasses():
+    num = random.randint(1, 10)
```
Comment (m-vdb) [inline tests/shared/utils/test_common.py:7]: I'm not sure if the use of `random` here is useful, at all

File: rasa/train.py (hunk 4 above)
Comment (wochinge) [reply rasa/train.py:15]: I'll leave these as they are as this is going beyond the scope of this PR. Will change all imports from `rasa.shared` though which aren't imported by module.

File: tests/shared/utils/test_common.py [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+    class TestClass:
+        pass
+
+    classes = [type(f"TestClass{i}", (TestClass,), {}) for i in range(num)]
```
Comment (m-vdb) [inline tests/shared/utils/test_common.py:12]: it doesn't really test for sub-subclasses and doesn't check that if a class is not a subclass, then it's not included

(i know you haven't introduced the code, feel free to add a FIXME if lack of time)

**Comment by wochinge** [comment]
@alwx Thanks for the review!  Implemented the changes

File: tests/shared/test_shared.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+            ]
+
+            # The shared package is required to be independent of the rest of Rasa
+            assert not outside_rasa_imports
```
Comment (m-vdb) [inline tests/shared/test_shared.py:30]: a more meaningful error message could be:
```suggestion
            assert not outside_rasa_imports, (
                f"File {file} imports code from outside "
                f"of `rasa.shared`: {','.join(outside_rasa_imports)}"
            )
```

And if there are several files that could raise errors, it might be nice to aggregate all the errors first and raise at the end with a summary of all files.

**Review by m-vdb — COMMENTED** [review]
Looks good! I added a few comments.

One other thing that comes to mind and challenges a bit the use of `rasa.shared.utils` from inside Rasa: if we move things to `rasa.shared` because they constitute an interface for `rasa`, it might not make sense for `rasa` to import utils from there (mirroring what we decided for rasa x). I know it may sound like a lot of duplication (for instance 3 `raise_warning` functions).

So I would suggest: either decide that `raise_warning` is part of this interface, and do not do duplication, or decide that it's not part of the interface, and never import it from `rasa.shared`. But not do it half-half. This decision doesn't need to be done here, and can be done in the end when we have a clearer picture of what's in utils. WDYT?

**Comment by wochinge** [comment]
> But not do it half-half.

By "half-half" you mean that Rasa X currently duplicates it and Rasa Open Source doesn't? I think we really have to see how this whole `utils` package develops when we are moving more things to `rasa.shared`. Currently I lean towards not having it duplicated (this would mean we can undo some of the duplication within Rasa X). So I agree to `this decision doesn't need to be done here, and can be done in the end when we have a clearer picture of what's in utils`  👍

File: tests/shared/test_shared.py (hunk 7 above)
Comment (wochinge) [reply tests/shared/test_shared.py:26]: it unlikely that somebody does `from rasa import shared` but added it 👍

File: tests/shared/utils/test_common.py (hunk 9 above)
Comment (wochinge) [reply tests/shared/utils/test_common.py:12]: > it doesn't really test for sub-subclasses 

It does actually create subclasses (https://docs.python.org/3/library/functions.html#type)

>  doesn't check that if a class is not a subclass, then it's not included

What would you like to test here? The fact that other, random classes, aren't found by this function?

**Review by alwx — APPROVED** [review]
🚀🚀🚀

File: tests/shared/utils/test_common.py (hunk 9 above)
Comment (m-vdb) [reply tests/shared/utils/test_common.py]: I meant sub-sub classes so that uses the recursion

Comment (wochinge) [reply tests/shared/utils/test_common.py]: will address that in a separate PR 👍 (wanted to get this merged)

**Comment by m-vdb** [comment]
@wochinge @alwx I think we forgot about the [`CODEOWNERS` file](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners) :octocat:

**Comment by wochinge** [comment]
will add it here https://github.com/RasaHQ/rasa/pull/6576
